hwp_gripper/control/pru_monitor.py

```diff
@@ -0,0 +1,291 @@
+        """Returns True if the gripper state has not been updated in a while"""
+        return time.time() - self.last_packet_received > self._expiration_time
+
+    def update(self, packet):
+        """Updates the gripper state based on an incoming PRU packet"""
+        self.last_packet_received = packet.timestamp
+
+        if isinstance(packet, EncoderPacket):
+            # Update the position of each encoder
+            for act in self.actuators:
+                idx = act.axis - 1
+                a_state = (packet.state >> 2*idx) & 1
+                b_state = (packet.state >> (2*idx + 1)) & 1
+
+                if self._last_enc_state is not None:
+                    rising_edges = np.diff(
+                        a_state, prepend=(self._last_enc_state >> 2*idx) & 1
+                    ) == 1
+                    dirs = (b_state[rising_edges] * 2 - 1)
+                else:
+                    rising_edges = np.diff(a_state) == 1
+                    dirs = (b_state[1:][rising_edges] * 2 - 1)
+
+                act.pos += np.sum(dirs) * MM_PER_NOTCH
+
+            self._last_enc_state = packet.state[-1]
+
+        elif isinstance(packet, LimitPacket):
+            self.last_limit_received = packet.timestamp
+
+            # Update the limit state of each actuator
+            for act in self.actuators:
+                for limit in act.limits.values():
+                    limit.state = bool((packet.state >> limit.pru_bit) & 1)
+
+        elif isinstance(packet, ErrorPacket):
+            # Should we do anything else for errors?
+            print(f"Error packet received!! Error code: {packet.error_code}")
+
+        elif isinstance(packet, TimeoutPacket):
+            # Should we do anything else for timeouts?
+            print(
+                f"Timeout packet received!! timeout type: {packet.timeout_type}")
+
+        else:
+            raise RuntimeError(f"Unknown packet type: {packet}")
+
+    def set_home(self):
+        for act in self.actuators:
+            act.pos = 0
+            act.calibrated = True
+
+
+class PruMonitor:
+    """
+    This class monitors incoming UDP packets from the PRU, and uses incoming
+    data to keep track of the Gripper state.
+
+    Args
+    ------
+    port : int
+        Port to listen for incoming UDP packets
+    ip_address : str
+        IP Address to listen for incoming UDP packets. Defaults to using all
+        available interfaces.
+    """
+
+    def __init__(self, port, ip_address=''):
+        self.port = port
+        self.ip_address = ip_address
+
+        self._gripper_state = GripperState()
+        self._state_lock = threading.Lock()
+
+        self.socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+        self.socket.bind((self.ip_address, self.port))
+        self.packet_queue = queue.Queue()
+
+        self.read_thread = threading.Thread(target=self._read_packets)
+        self.update_thread = threading.Thread(target=self._update_state)
```

Comment:
I'm not too familiar with the threading package. How well does this code handle unexpected exits? Will the threads close gracefully or will they continue running?